<!doctype html>
<html>
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>RAST source code scanning summary</title>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<link rel="stylesheet" href="css/bootstrap.min.css">
<script src="js/jquery-2.1.4.min.js" type="text/javascript"></script>
<script src="js/bootstrap.min.js"></script>
<link href="css/jquerysctipttop.css" rel="stylesheet" type="text/css">
<!-- INCLUDES -->
<link rel="stylesheet" href="css/bootstrap-table-expandable.css">
<script src="js/bootstrap-table-expandable.js"></script>
<style>
h2 {color: #7c795d; font-family: 'Trocchi', serif; font-size: 45px; font-weight: normal; line-height: 48px; margin: 0; text-align: center;   }
table.repo_table {
color:#fff;
font-family:Helvetica,Arial,sans-serif;
border-collapse: collapse;
padding:0 3px;
border-radius: 5px;
margin: 25px auto;
}

table.repo_table td, table.repo_table th {
height:30px;
transition:all 1.5s;
text-align:center;
padding: 5px 40px;
border: 2px solid #000;
width: 200px;
}

table.repo_table th {
background:#000;
font-weight:700;
}

table.repo_table tr td {
background:#fff;
color:#000;
}

table.repo_table tr td:hover {
background:#666;
color:#FFF;
font-weight:bold;
}
</style>
</head>
<body>



<div class="container">
<br>
<br>
<br>
<br>
  <h2>RAST Summary</h2>

<!--  <p><strong>Note:</strong> The <strong>data-parent</strong> attribute makes sure that all collapsible elements under the specified parent will be closed when one of the collapsible item is shown.</p>
<p>Demo for <a href="http://www.jankoatwarpspeed.com/post/2009/07/19/Expand-table-rows-with-jQuery-jExpand-plugin.aspx">Expand table rows with jQuery - jExpand plugin</a></p> -->

<table class="repo_table">
    <thead>
        <tr>
            <th>Rule Name</th>
            <th>Total Violations</th>
            
        </tr>
    </thead>
    <tbody>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/DCL00-J.+Prevent+class+initialization+cycles" target="_blank"> DCL00J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/DCL01-J.+Do+not+reuse+public+identifiers+from+the+Java+Standard+Library" target="_blank"> DCL01-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/DCL02-J.+Do+not+modify+the+collection%27s+elements+during+an+enhanced+for+statement" target="_blank"> DCL02-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/EXP02-J.+Do+not+use+the+Object.equals%28%29+method+to+compare+two+arrays" target="_blank"> EXP02-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/ERR02-J.+Prevent+exceptions+while+logging+data" target="_blank"> ERR02-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/ERR04-J.+Do+not+complete+abruptly+from+a+finally+block" target="_blank"> ERR04-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J</a></td>
<td>16</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank"> ERR07-J</a></td>
<td>28</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/ERR02-J.+Prevent+exceptions+while+logging+data" target="_blank"> ERR08-J</a></td>
<td>1</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/FIO02-J.+Detect+and+handle+file-related+errors" target="_blank"> FIO02-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/LCK01-J.+Do+not+synchronize+on+objects+that+may+be+reused" target="_blank"> LCK01-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/pages/viewpage.action?pageId=43647087" target="_blank"> LCK02-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/MET09-J.+Classes+that+define+an+equals%28%29+method+must+also+define+a+hashCode%28%29+method" target="_blank"> MET09-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/MSC00-J.+Use+SSLSocket+rather+than+Socket+for+secure+data+exchange" target="_blank"> MSC00-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/MSC01-J.+Do+not+use+an+empty+infinite+loop" target="_blank"> MSC01-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/MSC02-J.+Generate+strong+random+numbers" target="_blank"> MSC02-J</a></td>
<td>1</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/NUM07-J.+Do+not+attempt+comparisons+with+NaN" target="_blank"> NUM07-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/NUM09-J.+Do+not+use+floating-point+variables+as+loop+counters" target="_blank"> NUM09-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/NUM10-J.+Do+not+construct+BigDecimal+objects+from+floating-point+literals" target="_blank"> NUM10-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/OBJ09-J.+Compare+classes+and+not+class+names" target="_blank"> OBJ09-J</a></td>
<td>0</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/OBJ10-J.+Do+not+use+public+static+nonfinal+fields" target="_blank"> OBJ10-J</a></td>
<td>7</td>
 </tr>
<tr>
<td><a href="https://www.securecoding.cert.org/confluence/display/java/SER05-J.+Do+not+serialize+instances+of+inner+classes" target="_blank"> SER05-J</a></td>
<td>0</td>
 </tr>

 </tbody>
    </table> 
<br>
<hr noshade>
<br>
<h4 style="color:#660000;">Total 3 directories searched and 78 files scanned. Out of the above rules, 53 CERT violations are reported </h4>
<br>
<hr noshade>
<br>
<div style="text-align:center">  
     <button class="btn" data-toggle="collapse" data-target="#demo">View Detailed Report</button>
</div> 
<div id="demo" class="collapse">
<h4>Following files have secure coding violations </h4>
  <div class="panel-group" id="accordion">
    <div class="panel panel-default">
  
        <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc6">AppsCustomizePagedView.java</a>
        </h4>
      </div>
      <div id="Loc6" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/AppsCustomizePagedView.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/AppsCustomizePagedView.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>1128</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">1128 </span> A method invocation <span style="color: red">data.cleanup(true) </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>1326</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">1326 </span> A method invocation <span style="color: red">data.cleanup(false) </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc7">AppsCustomizeTabHost.java</a>
        </h4>
      </div>
      <div id="Loc7" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/AppsCustomizeTabHost.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/AppsCustomizeTabHost.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>458</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">458 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc11">CellLayout.java</a>
        </h4>
      </div>
      <div id="Loc11" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/CellLayout.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/CellLayout.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>984</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">984 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>3136</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">3136 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc15">LauncherProvider.java</a>
        </h4>
      </div>
      <div id="Loc15" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/LauncherProvider.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/LauncherProvider.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>136</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">136 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>176</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">176 </span> A method invocation <span style="color: red">db.endTransaction() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>423</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">423 </span> A method invocation <span style="color: red">cursor.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>491</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">491 </span> A method invocation <span style="color: red">db.endTransaction() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>515</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">515 </span> A method invocation <span style="color: red">db.endTransaction() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>546</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">546 </span> A method invocation <span style="color: red">db.endTransaction() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>626</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">626 </span> A method invocation <span style="color: red">db.endTransaction() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>703</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">703 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>761</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">761 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>773</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">773 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>793</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">793 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>878</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">878 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>950</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">950 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>1010</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">1010 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>1174</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">1174 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>1177</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">1177 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc17">LauncherViewPropertyAnimator.java</a>
        </h4>
      </div>
      <div id="Loc17" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/LauncherViewPropertyAnimator.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/LauncherViewPropertyAnimator.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>76</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">76 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>81</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">81 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>179</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">179 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc20">PagedViewCellLayout.java</a>
        </h4>
      </div>
      <div id="Loc20" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/PagedViewCellLayout.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/PagedViewCellLayout.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>178</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">178 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc21">PagedViewCellLayoutChildren.java</a>
        </h4>
      </div>
      <div id="Loc21" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/PagedViewCellLayoutChildren.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/PagedViewCellLayoutChildren.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>86</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">86 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc33">HolographicOutlineHelper.java</a>
        </h4>
      </div>
      <div id="Loc33" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/HolographicOutlineHelper.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/HolographicOutlineHelper.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>134</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">134 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>163</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">163 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc42">UninstallShortcutReceiver.java</a>
        </h4>
      </div>
      <div id="Loc42" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/UninstallShortcutReceiver.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/UninstallShortcutReceiver.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>129</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">129 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc44">Utilities.java</a>
        </h4>
      </div>
      <div id="Loc44" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/Utilities.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/Utilities.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>174</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">174 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>MSC02-J</td>
		<td>CERT</td>
		<td>273</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>MSC02-J. Generate strong random numbers</h4>
     <p>When software generates predictable values in a context requiring unpredictability, it may be possible for an attacker to guess the next value that will be generated, and use this guess to impersonate another user or access sensitive information.

As the java.util.Random class relies on a pseudorandom number generator, this class and relating java.lang.Math.random() method should not be used for security-critical applications or for protecting sensitive data. In such context, the java.security.SecureRandom class which relies on a cryptographically strong random number generator (RNG) should be used in place.</p>
        
        <p>In the given program line number <span style="color: red">273 </span> code block <span style="color: red">new Random(System.currentTimeMillis()) </span> trying to </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p>This noncompliant code example uses the insecure java.util.Random class. This class produces an identical sequence of numbers for each given seed value; consequently, the sequence of numbers is predictable. </p>
 <pre><code class="language-java"> 
 
import java.util.Random;
// ...
 
Random number = new Random(123L);
//...
for (int i = 0; i < 20; i++) {
  // Generate another random integer in the range [0, 20]
  int n = number.nextInt(21);
  System.out.println(n);
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>This compliant solution uses the java.security.SecureRandom class to produce high-quality random numbers:</p>
 <pre><code class="language-java"> 
 
import java.security.SecureRandom;
import java.security.NoSuchAlgorithmException;
// ...
 
public static void main (String args[]) {
  SecureRandom number = new SecureRandom();
  // Generate 20 integers 0..20
  for (int i = 0; i < 20; i++) {
    System.out.println(number.nextInt(21));
  }
}
 </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="" target="_blank">MSC02-J. Generate strong random numbers</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2006-6969" target="_blank">CVE-2006-6969</a> </li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc47">WidgetPreviewLoader.java</a>
        </h4>
      </div>
      <div id="Loc47" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/WidgetPreviewLoader.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/WidgetPreviewLoader.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>248</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">248 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>281</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">281 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>447</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">447 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>650</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">650 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc48">Workspace.java</a>
        </h4>
      </div>
      <div id="Loc48" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/Workspace.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/Workspace.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>2766</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">2766 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc51">LauncherModel.java</a>
        </h4>
      </div>
      <div id="Loc51" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/LauncherModel.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/LauncherModel.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>229</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">229 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>509</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">509 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>557</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">557 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>599</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">599 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>1082</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">1082 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>1088</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">1088 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>1094</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">1094 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
	  <tr>
		<td>ERR05-J</td>
		<td>CERT</td>
		<td>1504</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR05-J. Do not let checked exceptions escape from a finally block</h4>
     <p>Methods invoked from within a finally block can throw an exception. Failure to catch and handle such exceptions results in the abrupt termination of the entire try block. Abrupt termination causes any exception thrown in the try block to be lost, preventing any possible recovery method from handling that specific problem. </p>
        
        <p>In the given program line number <span style="color: red">1504 </span> A method invocation <span style="color: red">c.close() </span> is present inside a finally block without a try-catch block </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> This noncompliant code example contains a finally block that closes the reader object. The programmer incorrectly assumes that the statements in the finally block cannot throw exceptions and consequently fails to appropriately handle any exception that may arise. </p>
 <pre><code class="language-java"> 
 try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        reader.close();
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution encloses the close() method invocation in a try-catch block of its own within the finally block. Consequently, the potential IOException can be handled without allowing it to propagate further.t</p>
 <pre><code class="language-java"> 
  try {
      BufferedReader reader =
          new BufferedReader(new FileReader(some_file));
      try {
        // Do operations
      } finally {
        try {
          reader.close();
        } catch (IOException ie) {
          // Forward to handler
        }
        // ... Other cleanup code ...
      }
    } catch (IOException x) {
      // Forward to handler
    }
    </code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR05-J.+Do+not+let+checked+exceptions+escape+from+a+finally+block" target="_blank"> ERR05-J. Do not let checked exceptions escape from a finally block</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/703.html" target="_blank"> CWE-703,</a> Improper Check or Handling of Exceptional Conditions</li>
          <li><a href="http://cwe.mitre.org/data/definitions/248.html" target="_blank"> CWE-248,</a> Uncaught Exception</li>
          
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
  <div class="panel panel-default">
      <div class="panel-heading">
        <h4 class="panel-title">
          <a data-toggle="collapse" data-parent="#accordion" href="#Loc72">Launcher.java</a>
        </h4>
      </div>
      <div id="Loc72" class="panel-collapse collapse collapse">
        <div class="panel-body"> <table class="table table-hover table-expandable table-striped">
    <thead>
 <p>Location :<a href="/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/Launcher.java">/media/rahul/Education/Project/src/M/Sources/packages/apps/Launcher2/src/com/android/launcher2/Launcher.java</a></p>

      <tr>
        <th>Rule</th>
        <th>Standard</th>
        <th>Line number</th>
        
      </tr>
    </thead>
    <tbody>
      
      
   	  <tr>
		<td>ERR08-J</td>
		<td>CERT</td>
		<td>1623</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR08-J. Do not catch NullPointerException or any of its ancestors</h4>
     <p>NullPointerException should be avoided, not caught. Any situation in which NullPointerException is explicitly caught can easily be converted to a null test, and any behavior being carried out in the catch block can easily be moved to the "is null" branch of the conditional.</p>
        
        <p>In the given program line number <span style="color: red">1623 </span> catch block contains a null pointer exception <span style="color: red">NullPointerException ex </span>  </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p>In the Non-compliant Solution rather than checking to see whether the given string is null, the method catches NullPointerException and returns false. </p>
 <pre><code class="language-java"> 
public int lengthPlus(String str) {
  int len = 2;
  try {
    len += str.length();
  }
  catch (NullPointerException e) {
    log.info("argument was null");
  }
  return len;
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p>  This compliant solution explicitly checks the String argument for null rather than catching NullPointerException:</p>
 <pre><code class="language-java"> 
 public int lengthPlus(String str) {
  int len = 2;

  if (str != null) {
    len += str.length();
  }
  else {
    log.info("argument was null");
  }
  return len;
}
</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR08-J.+Do+not+catch+NullPointerException+or+any+of+its+ancestors" target="_blank"> ERR08-J. Do not catch NullPointerException or any of its ancestors</a></li>
          </ul> 
          </td>  </tr>
	  <tr>
		<td>ERR07-J</td>
		<td>CERT</td>
		<td>2672</td>
		<td><!--$001violation$001--></td>
	      </tr>
      

<tr>

      
        <td colspan="5">
 <h4>ERR07-J. Do not throw RuntimeException, Exception, or Throwable</h4>
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
     <p>Methods must not throw RuntimeException, Exception, or Throwable. Throwing a RuntimeException can lead to subtle errors; for example, a caller cannot examine the exception to determine why it was thrown and consequently cannot attempt recovery.
</p>
        
        <p>In the given program line number <span style="color: red">2672 </span> throws <span style="color: red">RuntimeException </span> </p> 
        <span style="color: #660000">  <h4>Non-compliant Solution</h4></span>
<p> The isCapitalized() method in this noncompliant code example accepts a string and returns true when the string consists of a capital letter followed by lowercase letters. The method also throws a RuntimeException when passed a null string argument. </p>
 <pre><code class="language-java"> 
boolean isCapitalized(String s) {
  if (s == null) {
    throw new RuntimeException("Null String");
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}
 </code></pre>
<span style="color:#00b33c"><h4>Compliant Solution</h4></span>
<p> This compliant solution throws NullPointerException to denote the specific exceptional condition</p>
 <pre><code class="language-java">
  boolean isCapitalized(String s) {
  if (s == null) {
    throw new NullPointerException();
  }
  if (s.equals("")) {
    return true;
  }
  String first = s.substring(0, 1);
  String rest = s.substring(1);
  return (first.equals(first.toUpperCase()) &&
          rest.equals(rest.toLowerCase()));
}</code></pre>
<h4>Additional information</h4>
          <ul>
            <li><a href="https://www.securecoding.cert.org/confluence/display/java/ERR07-J.+Do+not+throw+RuntimeException%2C+Exception%2C+or+Throwable" target="_blank">ERR07-J. Do not throw RuntimeException, Exception, or Throwable</a></li>
          </ul> 
          <h4>Related Guidelines</h4>
          <ul>
            <li><a href="http://cwe.mitre.org/data/definitions/397.html" target="_blank">CWE-397</a>, Declaration of Throws for Generic Exception</li>
          </ul></td>  </tr>
<!--$newtableblockattachhere-->
    </tbody>
  </table></div>
      </div>
    </div>

    
 
 <!--$newfileblockattachhere-->
    
  </div> 
</div>
</div>   
<br>
<br>
<br>
<br>
<br>
</body>
</html>

